### PR TITLE
Details of wrapInScript/Roddy interface.

### PR DESCRIPTION
**Pull request by vinjana**
Added diagram of control flow in `wrapInScript.sh`. Describe in more detail the interface between Roddy and the wrapper.

## Summary by CodeRabbit

- **Documentation**
  - Updated `README.md` with detailed descriptions of the `wrapInScript.sh` wrapper script for bioinformatic job execution on a cluster.
  - Clarified script purpose, calling conventions, file creation, environment setup support, and wrapped-script execution.
  - Expanded details on job-state logfiles, lock-files, extended logs, environment scripts, and debugging options.

- **New Features**
  - Introduced `wrapInScript.sh` wrapper script for streamlined bioinformatic job execution.
  - Added environment variables and support for job-specific environment scripts.

- **New Diagrams**
  - Added sequence, component, and state diagrams to illustrate the flow and relationships in the system.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,41 +2,72 @@
...
 
-All top-level tools or scripts that are supposed to be started on the cluster by Roddy are actually not directly started, but are wrapped by the `resources/roddyTools/wrapInScript.sh` contained in this plugin.
+The overall structure of Roddy and its plugins can be depicted like this:
+
+![image](images/components.svg)
+
+The [PluginBase](https://github.com/TheRoddyWMS/Roddy-Base-Plugin) provides a base class `BasePlugin` for the JVM-compiled code of all plugins.
+Roddy uses this JVM-compiled code to manage cluster jobs and do submission-time checks of input files (usually by accessing the files from the cluster's head-node).
+
+By contrast, the DefaultPlugin provides a wrapper script `wrapInScript.sh` which is used to wrap the top-level, bioinformatic (or "wrapped") script that are developed by workflow developers for each (cluster) job.
```
Comment (GWarsow) [inline README.md]: script**s** that are developed
or
script that **is** developed

File: README.md [hunk 2]
```diff
@@ -2,41 +2,72 @@
...
+
+## Calling Conventions of the `wrapInScript.sh`
+
+The `wrapInScript.sh` is called without any parameters. 
+The entire configuration of the `wrapInScript.sh` and the wrapped top-level bioinformatic script is done via environment variables and a configuration file.
+
+* `PARAMETER_FILE` is the path to the parameter file that is loaded (and thus evaluated) by the Bash `source` command.
+The parameter file should `export` all variables and `export -f` all functions that are needed by the wrapped script.
+* `debugWrapInScript` is a string variable that turns on debugging options in the `wrapInScript.sh`.
+* `outputFileGroup`: Can be set to the name of existing group of which the used user is a member. This group will be used for the output files of the wrapped script.
```
Comment (GWarsow) [inline README.md]: to the name of **an** existing **unix** group of which the ~~used~~ user

File: README.md [hunk 3]
```diff
@@ -2,41 +2,72 @@
...
-  - source job-specific environment script (see "Environment Setup Support" below)
-  - setup scratch directory
-  - update the `jobStateLogfile.txt` using a lock-file
-  - run the wrapped script using bash 
-  - kill child-process still running after the wrapped script ended
-  - update the `jobStateLogfile.txt` with the job's exit code
-  - exit
+* job-state logfile:
+* job-state logfile lock-file: This has the same name and location as the job-state logfile, but with the suffix `~` appended.
+The lock is created only twice, before the job-state logfile is accessed, and deleted after each write operation was finished.
```
Comment (GWarsow) [inline README.md]: I do not understand the logic behind `lock is created only twice`

File: README.md [hunk 4]
```diff
@@ -2,41 +2,72 @@
...
+* `extendedLogs` directory: This directory contains the extended logs of the wrapped script and creates a file for each cluster job (using the job's parameter file name as template).
+It contains two dumps of the environment during the execution of the wrapped script -- before and after the parameter file was `source`d.
```
Comment (GWarsow) [inline README.md:51]: Could you provide the name/pattern of those files?

File: README.md [hunk 5]
```diff
@@ -2,41 +2,72 @@
...
 
 ## Environment Setup Support
 
 ### Base Environment Script
 
-Each job is started with the default environment configured in you `applicationProperties.ini` in the `baseEnvironmentScript` variable. The `baseEnvironmentScript` serves as kind of general configuration of your cluster environment. Usually you will use a script like `/etc/profile` or `$HOME/.profile` or `$HOME/.bashrc`. 
+Each job is started with the default environment configured in your `applicationProperties.ini` in the `baseEnvironmentScript` variable.
```
Comment (GWarsow) [inline README.md]: in the `baseEnvironmentScript` variable of your `applicationProperties.ini` file.

File: README.md [hunk 6]
```diff
@@ -2,41 +2,72 @@
...
+The `baseEnvironmentScript` serves as kind of general configuration of your cluster environment.
```
Comment (GWarsow) [inline README.md]: as **a**(?) kind of general

File: README.md [hunk 7]
```diff
@@ -64,23 +99,28 @@ You may want to specify dedicated job-environment scripts for individual cluster
...
   - the value contains a '/': this is a direct path. This only works since version 1.2.2-5.
   - the value starts with '${': this is a TOOL_ path. Since version 1.2.2-5 the matching is on `${TOOL_}`.
-  - compose the the `TOOL_` variable name from the job-name, like described above.
+  - compose the `TOOL_` variable name from the job-name, like described above.
 
 ### Environment Parametrization
 
-The environment script is simply `source`'d, so you can access variables from the parameter-file (`PARAMETER_FILE`, sourced before; see above) from within that script. For instance, you have a `conda.sh` that activates a Conda environment, but you want to keep the environment name configurable. You can then the conda environment name in the XML:
+The environment script is simply `source`'d, so you can access variables from the parameter-file (`PARAMETER_FILE`, sourced before; see above) from within that script. For instance, you have a `conda.sh` that activates a Conda environment, but you want to keep the environment name configurable. 
+You can then the conda environment name in the XML:
```
Comment (GWarsow) [inline README.md]: some word seems to be missing here

File: README.md [hunk 8]
```diff
@@ -99,9 +139,12 @@ The environment setup scripts are mostly useful for setting up environment varia
 
 To achieve this Bash variables need to be exported with the `export` declaration.
```
Comment (GWarsow) [inline README.md]: I would suggest to add a comma after `To achieve this`

File: README.md [hunk 9]
```diff
@@ -119,15 +162,20 @@ The `debugWrapInScript` variable -- defaulting to `false` -- turns on the `set +
   
 The `baseEnvironmentScript` is sourced with relaxed values for `set`, i.e. with `set +ue`, because often files like `/etc/profile` are not under the control of the person running the workflow. Conversely, changes to the `set` options in the `baseEnvironmentScript` are not inherited by subsequent code in the  `wrapInScript.sh`.
```
Comment (GWarsow) [inline README.md]: there is a doubled whitespace at the end of the sentence: `code in the  \`wrapInScript.sh\`.`

**Review by GWarsow — CHANGES_REQUESTED** [review]
the suggested changes make application and behavior or Roddy much clearer. I have only minor points to be mentioned.

File: README.md (hunk 9 above)
Comment (GWarsow) [reply README.md]: would you like to keep it?

**Review by GWarsow — APPROVED** [review]

**Comment by vinjana** [comment]
@GWarsow Thanks!

**Pull request merged**